module/zfs/zfs_ioctl.c

```diff
@@ -284,6 +285,65 @@ static int zfs_fill_zplprops_root(uint64_t, nvlist_t *, nvlist_t *,
 int zfs_set_prop_nvlist(const char *, zprop_source_t, nvlist_t *, nvlist_t *);
 static int get_nvlist(uint64_t nvl, uint64_t size, int iflag, nvlist_t **nvp);
 
+/* Returns 1 on failure; Returns 0 on success */
+int fetch_zpool_stats(zpool_size_stats* zstats) {
+    int ret = 1;
+
+    // If 'zpool/zvol' name is provided, truncating it to zpool name only
+    char* slash = strchr(zstats->zpool_name, '/');
+    if (slash)
+	*slash = '\0';
+
+    mutex_enter(&spa_namespace_lock);
+    spa_t* spa;
+    if ((spa = spa_lookup(zstats->zpool_name)) != NULL) {
+	mutex_enter(&spa->spa_props_lock);
+	metaslab_class_t* mc = spa_normal_class(spa);
+
+	zstats->alloc = metaslab_class_get_alloc(mc);
+	zstats->alloc += metaslab_class_get_alloc(spa_special_class(spa));
+	zstats->alloc += metaslab_class_get_alloc(spa_dedup_class(spa));
+
+	zstats->size = metaslab_class_get_space(mc);
+	zstats->size += metaslab_class_get_space(spa_special_class(spa));
+	zstats->size += metaslab_class_get_space(spa_dedup_class(spa));
+	mutex_exit(&spa->spa_props_lock);
+	ret = 0;
+    }
+    mutex_exit(&spa_namespace_lock);
+    return ret;
+}
+
+extern zvol_state_t* wzvol_find_target(uint8_t targetid, uint8_t lun);
+extern void wzvol_unlock_target(zvol_state_t* zv);
+
+NTSTATUS zpool_get_size_stats(PDEVICE_OBJECT DeviceObject, PIRP Irp, PIO_STACK_LOCATION IrpSp)
```

Comment:
Adding Windows specific things like PDEVICE_OBJECT into the shared code could be awkward. We usually define our own ioctl functions which live in `zfs_ioctl_os.c`. 
https://github.com/openzfsonwindows/openzfs/blob/windows/module/os/windows/zfs/zfs_ioctl_os.c#L263-L267

Reply:
Removed the non-DI changes and shall add in a separate PR.